cpp/src/arrow/compute/kernels/scalar_fill_null.cc

```diff
@@ -32,195 +34,117 @@ namespace internal {
 
 namespace {
 
-template <typename OutType, typename InType, typename Enable = void>
+template <typename Type, typename Enable = void>
 struct FillNullFunctor {};
 
-template <typename OutType, typename InType>
-struct FillNullFunctor<OutType, InType, enable_if_t<is_number_type<InType>::value>> {
-  using value_type = typename TypeTraits<InType>::CType;
-  using BuilderType = typename TypeTraits<OutType>::BuilderType;
+template <typename Type>
+struct FillNullFunctor<Type, enable_if_t<is_number_type<Type>::value>> {
+  using T = typename TypeTraits<Type>::CType;
 
   static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
-    const Datum& in_arr = batch[0];
-    const Datum& fill_value = batch[1];
-
-    if (!in_arr.is_arraylike()) {
-      ctx->SetStatus(Status::Invalid("Values must be Array or ChunkedArray"));
-    }
-    if (!fill_value.is_scalar()) {
-      ctx->SetStatus(Status::Invalid("fill value must be a scalar"));
-    }
```

Comment:
Note: these type checks are not necessary. You can safely assume that once `Exec` is called that the types have already been checked